common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp

```diff
@@ -507,6 +552,154 @@ bool base_wlan_hal_dwpal::attach_ctrl_interface(int vap_id)
+                                 NL80211_CMD_VENDOR, DWPAL_NETDEV_ID,
+                                 (enum ltq_nl80211_vendor_subcmds)nl_cmd, vendor_data,
+                                 vendor_data_size) != DWPAL_SUCCESS) {
+        err = false;
+    }
+
+    return err;
+}
+
+bool base_wlan_hal_dwpal::dwpal_nl_cmd_get(const std::string &ifname, unsigned int nl_cmd,
+                                           unsigned char *out_buffer, size_t *out_buffer_size,
+                                           const size_t max_buffer_size)
+{
+
+    if ((out_buffer_size == NULL) || (out_buffer == NULL)) {
+        LOG(ERROR) << __FUNCTION__ << "out_buffer or out_buffer_size incorrect ==> Abort!\n";
+        return false;
+    }
+
+    /* Handle a command which invokes an event with the output data */
+    if (dwpal_driver_nl_cmd_send(
+            m_dwpal_nl_ctx, DWPAL_NL_SOLICITED_EVENT, (char *)ifname.c_str(), NL80211_CMD_VENDOR,
+            DWPAL_NETDEV_ID, (enum ltq_nl80211_vendor_subcmds)nl_cmd, NULL, 0) == DWPAL_FAILURE) {
+        LOG(DEBUG) << __FUNCTION__ << "ERROR for cmd = " << nl_cmd;
+        return false;
+    }
+
+    // Passing a lambda with capture is not supported for standard C function
+    // pointers. As a workaround, we create a static (but thread local) wrapper
+    // function that calls the capturing lambda function.
+    static __thread std::function<DWPAL_Ret(char *ifname, int event, int subevent, size_t len,
+                                            unsigned char *data)>
+        nl_handler_cb_wrapper;
+    nl_handler_cb_wrapper = [&](char *ifname, int event, int subevent, size_t len,
+                                unsigned char *data) -> DWPAL_Ret {
+        LOG(DEBUG) << __func__ << " ifname = " << std::string(ifname) << " event = " << event
+                   << " subevent = " << subevent << " len = " << len;
+
+        if ((len == 0) || (data == NULL)) {
+            LOG(DEBUG) << __func__ << "len=0 and/or data is NULL ==> Abort!\n";
+            return DWPAL_FAILURE;
+        }
+        if (event == NL80211_CMD_VENDOR) {
+            if (len >= max_buffer_size) {
+                LOG(DEBUG) << __func__ << "NL size exceeds out_buffer size ==> Abort!\n";
+                return DWPAL_FAILURE;
+            }
+            // copy get result from nl buffer (data[i])
+            for (int i = 0; i < (int)len; i++) {
+                out_buffer[i] = data[i];
+            }
+            *out_buffer_size = len;
+
+        } else {
+            LOG(DEBUG) << __func__ << "non vendor event not handled ";
+            return DWPAL_FAILURE;
+        }
+        return DWPAL_SUCCESS;
+    };
+    auto nl_handler_cb = [](char *ifname, int event, int subevent, size_t len,
+                            unsigned char *data) -> DWPAL_Ret {
+        return nl_handler_cb_wrapper(ifname, event, subevent, len, data);
+    };
+
+    while (*out_buffer_size == 0) { /* wait for reply with valid data */
```

Comment:
I think it makes much more sense to add a promise or something like this instead of this (ugly) hack.

Reply:
Doing a select()/poll() on the appropriate fd should work.

Reply:
its ported, we can try to remove it at validation stage